### PR TITLE
Extend buffer donation aliasing APIs

### PR DESCRIPTION
**Pull request by rpsilva-aws**
Enables https://github.com/pytorch/xla/issues/8711. In particular, it alleviates the fallback aliasing logic for user config. Previously, it would only be applied if auto LTC did not infer any needed aliasing which is a semantically suboptimal surface area. In this PR, we modify this behavior to always accommodate explicitly donated buffers, working simultaneously with LTC, IF enabled.

In this PR, it is out of scope to address the per-device concerns associated with having global (DeviceArenaContext) metadata for all the devices with respect to aliasing and other dynamo execution -specific logic.

### CODE REVIEW

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 1]
```diff
@@ -1287,56 +1297,70 @@ std::vector<size_t> GetBufferDonorIndexForStepMarker(
...
-      // But, when we issue a step barrier (force_ltc_data == true) we have to
-      // turn everything into DEVICE_DATA, so we can activate aliasing.
-      buffer_donor_indices = GetBufferDonorIndexForStepMarker(
-          tensors, coll.indices, parameters_data);
-    } else if (GetAliasWithBufferDonorConfig()) {
-      // only alias based on buffer donor if LTC can't auto infer the input
-      // output aliasing.
-      buffer_donor_indices = GetBufferDonorIndexFromUserConfig(parameters_data);
-    }
+  if (ShardingUtil::GetAutoSharding()) {
```
Comment (ysiraichi) [inline torch_xla/csrc/xla_graph_executor.cpp:1303]: Could we move the refactoring in another PR?

**Review by ysiraichi — COMMENTED** [review]
Overall, LGTM. I do think it would be better to leave non-trivial refactorings for another PR. It would make reviewing easier.

That said, I'm not entirely familiar with this buffer donor system. Maybe @tengyifei @lsy323 @vanbasten23 @bhavya01 could help review.

File: test/dynamo/test_dynamo_aliasing.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import sys
```
[the remark this answers is not in the document]
Comment (rpsilva-aws) [reply test/dynamo/test_dynamo_aliasing.py:1]: It is used, but the import was missing.

File: test/test_input_output_aliases.py [hunk 3]
```diff
@@ -210,7 +239,104 @@ def test_device_data_cache_no_aliasing(self):
...
+  test_cases = loader.getTestCaseNames(InputOutputAliasesTest)
+  failed = False
+  for test_name in test_cases:
+    test = InputOutputAliasesTest(test_name)
+    runner = unittest.TextTestRunner(failfast=True)
+    result = runner.run(test)
+    if not result.wasSuccessful():
+      failed = True
+
+  sys.exit(1 if failed else 0)
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply test/test_input_output_aliases.py:343]: Do you mean that we would have tests running in parallel? As far as I understand it, the tests are run in sequence, aren't they?

File: torch_xla/csrc/init_python_bindings.cpp [hunk 4]
```diff
@@ -2737,36 +2753,43 @@ void InitXlaModuleBindings(py::module m) {
...
-            data->set_should_donate_buffer(should_donate);
-            buffer_donation_updated = true;
-          } else if (xtensor->CurrentIrValue().node != nullptr) {
-            torch::lazy::NodePtr node = xtensor->CurrentIrValue().node;
-            auto device_data = torch_xla::DeviceData::Cast(node.get());
-            if (device_data != nullptr) {
-              device_data->set_buffer_donation(should_donate);
-              buffer_donation_updated = true;
+        [](const std::vector<at::Tensor>& tensors,
+           bool should_donate) -> std::vector<bool> {
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply torch_xla/csrc/init_python_bindings.cpp:2766]: Do you mean that in these real world examples this modification would improve performance in a non-negligible way? If so, I think I would want to see some kind of profiling, first.

File: torch_xla/csrc/xla_graph_executor.h [hunk 5]
```diff
@@ -276,7 +289,7 @@ class XLAGraphExecutor : public torch::lazy::LazyGraphExecutor {
     torch::lazy::Value IrValueFromScalar(
         const at::Scalar& value, at::ScalarType scalar_type,
         const torch::lazy::BackendDevice& device) final;
-    bool should_alias_with_buffer_donor = false;
+    BufferDonorAliasConfig buffer_donor_alias_config{};
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply torch_xla/csrc/xla_graph_executor.h:292]: That makes sense. We can consider adding more when the need arises, but for now, managing 2 fields separately still seems reasonable. (it also makes reviewing easier)

File: test/test_input_output_aliases.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rpsilva-aws) [reply test/test_input_output_aliases.py:343]: I thought so too, but I see lingering graph executor states across tests. This at least enforces that we treat each test fn as a suite, so I don't think it's a functional concern to switch it to this. We have the same complications when trying to use env vars (on/off) in the same test file, hence why most tests ended up to just specifying the env var prior to running the test (e.g. https://github.com/pytorch/xla/blob/master/test/run_tests.sh#L155) which I wanted to avoid doing here.

File: torch_xla/csrc/init_python_bindings.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rpsilva-aws) [reply torch_xla/csrc/init_python_bindings.cpp:2766]: We can, but from a user interface point alone, why do you think it's easier to call it multiple times? Especially considering that the use case is often over a set of tensors (deciding to donate gradients, weights, optimizer states, etc). I think this is a more general question, since we could say the same about any binding that takes in a vector of tensors. We wouldn't need to make this change in this PR and collect more evidence if you're hesitant to doing this change.

File: test/test_input_output_aliases.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ysiraichi) [reply test/test_input_output_aliases.py:343]: > I see lingering graph executor states across tests.

I think they do linger because we don't reset them in the `tearDown()` method.

> I don't think it's a functional concern to switch it to this.

I agree that it's not a functional concern. But I would say that having only this test with this custom runner might be a maintainability concern. Do you think we could keep the old state?

If you think we should adopt this strategy, I think it would be nice to have a separate issue so that people could comment in.

File: torch_xla/__init__.py [hunk 6]
```diff
@@ -195,6 +195,9 @@ def _check_deprecated_env_var():
 if os.environ.get('TF_CPP_MIN_LOG_LEVEL') == '0':
   logger.setLevel(logging.INFO)
 
+if 'XLA_ENABLE_PARAM_ALIASING' in os.environ:
+  _XLAC.set_enable_parameter_aliasing(os.environ['XLA_ENABLE_PARAM_ALIASING'])
+
```
[the remark this answers is not in the document]
Comment (rpsilva-aws) [reply torch_xla/__init__.py:200]: That's reasonable, I'll separate it.

File: test/test_input_output_aliases.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rpsilva-aws) [reply test/test_input_output_aliases.py:343]: I have also tried with setUp and tearDown to revert to the prior state, but didn't work out. I can do a bit more experimentation here. Unfortunately, this blocks this PR, but let me time box it.

File: torch_xla/csrc/init_python_bindings.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ysiraichi) [reply torch_xla/csrc/init_python_bindings.cpp:2766]: It's not that I'm hesitant to this change. It's just that this PR is introducing new behavior. So, I think it would be easier to review and understand it if this PR had only the set of changes for that.

This Python binding + the refactoring aren't necessary for this new behavior. They deserve their own issues (pitching that idea) and PRs.

That said, my opinion is that there seems not to be a compelling reason to do this Python binding change. However, I'm not really familiar with how it's used out in the wild. So, let's wait to hear what the others have to say.

[the remark this answers is not in the document]
Comment (rpsilva-aws) [reply torch_xla/csrc/init_python_bindings.cpp:2766]: Absolutely. Sounds good, let's separate it and I can come back to it with any potential impact data.

**Comment by rpsilva-aws** [comment]
As suggested by @ysiraichi, so we don't block on this needed PR, deferring:

> - Switch to using a batch call for the donation - since in practice, there should be a large number of tensors that can be collected and simultaneously annotated. Note that the dynamo execution API remain unchanged, as these are intentionally not modified in this PR.

In addition to separating the following as separate PRs (I'll do so shortly after closing this one):
- Refining the parameter disabling user behavior (https://github.com/pytorch/xla/pull/8721#discussion_r1964297682)
- Gradient accumulation API changes

**Comment by ysiraichi** [comment]
> I'll do so shortly after closing this one

Just so we are clear: are you closing this one? Or do you mean after this gets merged?

**Comment by rpsilva-aws** [comment]
> > I'll do so shortly after closing this one
> 
> Just so we are clear: are you closing this one? Or do you mean after this gets merged?

I will do these as separate PRs after this one is merged.

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 7]
```diff
@@ -1290,53 +1290,73 @@ std::vector<size_t> XLAGraphExecutor::GetBufferDonors(
...
+  // any duplicates when merging the two indices vector.
+  std::vector<size_t> buffer_donor_indices;
+  buffer_donor_indices.reserve(
+      std::max(ltc_buffer_donor_indices.size(),
+               user_config_buffer_donor_indices.size()));
+  std::set_union(ltc_buffer_donor_indices.cbegin(),
+                 ltc_buffer_donor_indices.cend(),
+                 user_config_buffer_donor_indices.cbegin(),
+                 user_config_buffer_donor_indices.cend(),
+                 std::back_inserter(buffer_donor_indices));
```
Comment (ysiraichi) [inline torch_xla/csrc/xla_graph_executor.cpp:1353]: Was this supposed to be a refactoring PR? Reason I'm asking is because these lines are changing the old behavior.

**Review by ysiraichi — COMMENTED** [review]
Just one comment. Looks good otherwise.
Could you also change the PR description so we know exactly what changes this PR is introducing?

File: torch_xla/csrc/xla_graph_executor.cpp (hunk 7 above)
Comment (rpsilva-aws) [reply torch_xla/csrc/xla_graph_executor.cpp:1353]: No, this is the actual PR that resolves the PFR. We're extending the buffer donation to interchangeably include user config buffer donations with LTC aliasing.

File: test/test_input_output_aliases.py [hunk 8]
```diff
@@ -210,6 +232,69 @@ def test_device_data_cache_no_aliasing(self):
...
+      t0 = torch.randn(4, 2, 2).to(xla_device)
+      t1 = torch.randn(4, 2, 2).to(xla_device)
+      self.assertTrue(torch_xla._XLAC._set_buffer_donation(t0, True))
+      self.assertTrue(torch_xla._XLAC._get_buffer_donation(t0))
+      self.assertFalse(torch_xla._XLAC._get_buffer_donation(t1))
+      t3 = t0 + t1
+      t1 += 2
+      xm.mark_step()
+
+      self.assertEqual(met.metric_data("InputOutputAliasCount")[1], 2.0)
```
Comment (tengyifei) [inline test/test_input_output_aliases.py:248]: Is it possible to test that t0 can no longer be used?

File: test/test_input_output_aliases.py [hunk 9]
```diff
@@ -210,6 +232,69 @@ def test_device_data_cache_no_aliasing(self):
...
+
+  def test_user_config_donation_with_ltc_donation_overlap(self):
+    with alias_with_buffer_donor_config_context(True):
```
Comment (tengyifei) [inline test/test_input_output_aliases.py:251]: What does this config context do?

File: test/test_input_output_aliases.py [hunk 10]
```diff
@@ -210,6 +232,69 @@ def test_device_data_cache_no_aliasing(self):
...
+      met.clear_all()
+      xla_device = xm.xla_device()
+      t0 = torch.randn(4, 2, 2).to(xla_device)
+      self.assertTrue(torch_xla._XLAC._set_buffer_donation(t0, True))
+      self.assertTrue(torch_xla._XLAC._get_buffer_donation(t0))
+      self.assertIn('XlaSetBufferDonation', met.counter_names())
+      self.assertEqual(met.counter_value('XlaSetBufferDonation'), 1)
+      t1 = t0 + 1
+      torch_xla._XLAC._xla_sync_multi([t0, t1], [str(xla_device)], True, False)
```
Comment (tengyifei) [inline test/test_input_output_aliases.py:272]: What does sync_multi do? Is it the same as mark_step?

File: torch_xla/csrc/xla_graph_executor.h [hunk 11]
```diff
@@ -95,8 +95,7 @@ class XLAGraphExecutor : public torch::lazy::LazyGraphExecutor {
   torch::lazy::BackendDataPtr GetBaseSeedData(
       const torch::lazy::BackendDevice& device);
 
-  void SetAliasWithBufferDonorConfig(bool should_alias);
-
+  void SetAliasWithBufferDonorConfig(bool enable_alias);
```
Comment (tengyifei) [inline torch_xla/csrc/xla_graph_executor.h:98]: I see you've renamed "should" to "enable". Does this imply some feature change?

**Review by tengyifei — COMMENTED** [review]
Apologies for the delay. I'll be frank I don't understand the internal details that well. In fact nobody on the team does. So I've asked some questions on the PR.

File: test/test_input_output_aliases.py (hunk 9 above)
Comment (rpsilva-aws) [reply test/test_input_output_aliases.py:251]: It enables the general behavior of buffer donation. In other words, this (`_xla_set_enable_alias_with_buffer_donor_config`) needs to be set, so that the tensors (with the associated XLA device data backend) that were explicitly marked to be donated (`_set_buffer_donation`) can be donated in the next execution (i.e. mark step) [1]. It's effectively a general opt-in for the entire donation behavior (arguably we wouldn't need it since `_set_buffer_donation` is already assuming that the user does intend to explicitly donate it, but I just kept it as is).

https://github.com/pytorch/xla/issues/8711 describes the problem and use case, but Jack introduced this in the context of dynamo execution (https://github.com/pytorch/xla/pull/6587). Since it's an opt-in, we want to give users the flexibility to explicitly annotate the donation (precisely when they don't need the tensor after mark step). This is particularly helpful with the gradient accumulation (and scan), since we can't do in-place mutations (aliasing is uncaptured by LTC), so this closes that allows us to mitigate / close that gap. The main change in this PR, to resolve the PFR, is that we enable LTC and buffer donation to coexist (and remove the previous odd constrained fallback behavior: if no LTC aliasings + user config is enabled -> donate the buffers).

[1] XLA HLO donation ref: https://github.com/openxla/xla/blob/1c7d5f62e3b4c90a4b3484429c5042c69ceb5c5b/xla/service/hlo.proto#L479

File: test/test_input_output_aliases.py (hunk 10 above)
Comment (rpsilva-aws) [reply test/test_input_output_aliases.py:272]: Mostly, but more flexible. It allows me to specify `sync_xla_data=False`, so that I can ensure that the aliasing is due to the user config (explicit buffer donation), and not auto LTC aliasings.

File: torch_xla/csrc/xla_graph_executor.h (hunk 11 above)
Comment (rpsilva-aws) [reply torch_xla/csrc/xla_graph_executor.h:98]: I understand using "should" for donating buffers (`_set_buffer_donation`), because it may not necessarily donate it, since it depends whether this config is enabled (`_xla_set_enable_alias_with_buffer_donor_config`). I renamed the parameters for `_xla_set_enable_alias_with_buffer_donor_config`/`SetAliasWithBufferDonorConfig`, since we are enabling/allowing explicitly marked tensors (`_set_buffer_donation`) to be donated, so I figured that renaming it to "enable" is more suitable.

**Comment by rpsilva-aws** [comment]
> Apologies for the delay. I'll be frank I don't understand the internal details that well. In fact nobody on the team does. So I've asked some questions on the PR.

No worries. Thanks for taking a look - let me know if I was able to answer these. I'll revise with the test change asap.

**Comment by rpsilva-aws** [comment]
@tengyifei Done, PTAL

**Comment by rpsilva-aws** [comment]
@tengyifei @ysiraichi Hey folks, can you PTAL.

File: test/test_input_output_aliases.py (hunk 9 above)
Comment (tengyifei) [reply test/test_input_output_aliases.py:254]: Got it. So we need both the config and an explicit donation call.

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 12]
```diff
@@ -1290,53 +1290,73 @@ std::vector<size_t> XLAGraphExecutor::GetBufferDonors(
...
+    // Also in the first example, unless we are doing a step barrier and hence
+    // include all live tensors, if the B value is not part of the graph, it
+    // will later fetch the new value of A, which is incorrect.
+    // But, when we issue a step barrier (force_ltc_data == true) we have to
+    // turn everything into DEVICE_DATA, so we can activate aliasing.
+    ltc_buffer_donor_indices = GetBufferDonorIndexForStepMarker(
+        tensors, coll.indices, parameters_data);
+  }
+
+  std::vector<size_t> user_config_buffer_donor_indices;
```
Comment (tengyifei) [inline torch_xla/csrc/xla_graph_executor.cpp:1334]: This "user config" name is really unfortunate. It should just be called "explicit donation". I was going to suggest renaming it until I realized this is an existing feature that you exposed.

**Review by tengyifei — APPROVED** [review]

File: torch_xla/csrc/xla_graph_executor.cpp (hunk 12 above)
Comment (rpsilva-aws) [reply torch_xla/csrc/xla_graph_executor.cpp:1334]: Yes, unfortunately these are tied to existing API bindings, so they could already be used by users. We could rename, but we'd have to do it over a few release cycles with a deprecation warning.

**Review by ysiraichi — APPROVED** [review]
LGTM.
